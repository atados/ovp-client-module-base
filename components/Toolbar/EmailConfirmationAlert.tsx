import useFetchAPIMutation from '~/hooks/use-fetch-api-mutation'
import React, { useState, useEffect } from 'react'
import Alert from '~/components/Alert'
import styled from 'styled-components'
import { Config } from '~/common'
import moment from 'moment'
import { useRouter } from 'next/router'

const AlertWrapper = styled.div`
  z-index: 51 !important;
`

export interface EmailConfirmationAlertProps {
  readonly user: any
}

const EmailConfirmationAlert: React.FC<EmailConfirmationAlertProps> = ({
  user,
}) => {
  const router = useRouter()
  const [showEmailConfirmationAlert, setShowEmailConfirmationAlert] = useState(
    false,
  )
  const mutation = useFetchAPIMutation<{ success: boolean }>(email => ({
    endpoint: '/users/request-email-verification/',
    method: 'POST',
    body: { email },
  }))

  useEffect(() => {
    const isConfigPage = router.pathname === '/eu/configuracoes'

    if (
      user?.uuid &&
      !user.is_email_verified &&
      (Config.emailConfirmation.warning || isConfigPage)
    ) {
      const storedString: string | null = localStorage.getItem(
        'hiddenEmailConfirmation',
      )

      if (storedString && !isConfigPage) {
        const data = JSON.parse(storedString)
        if (
          user.uuid === data.userUuid &&
          moment().isAfter(data.expirationDate)
        ) {
          setShowEmailConfirmationAlert(true)
        }
      } else {
        setShowEmailConfirmationAlert(true)
      }
    }
  }, [user?.uuid])

  const handleClose = () => {
    if (user?.uuid) {
      const hiddenEmailConfirmation = {
        userUuid: user.uuid,
        expirationDate: moment()
          .add(1, 'week')
          .format(),
      }

      localStorage.setItem(
        'hiddenEmailConfirmation',
        JSON.stringify(hiddenEmailConfirmation),
      )
    }
  }

  const handleClick = () => {
    if (user?.email) {
      mutation.mutate(user.email)
      handleClose()
    }
  }

  if (!showEmailConfirmationAlert) {
    return null
  }

  return (
    <AlertWrapper className="fixed left-0 bottom-0 right-0">
      <Alert onClose={handleClose}>
        {mutation.data?.success ? (
          'O email foi enviado para você, verifique na sua caixa de entrada.'
        ) : (
          <div>
            <b className="mr-1">Seu email ainda não foi confirmado.</b> Perdeu o
            email de confirmação?{' '}
            {mutation.loading ? (
              'Enviando...'
            ) : (
              <>
                <button
                  role="send-confirmation-email"
                  className="inline-block mx-1 underline"
                  onClick={handleClick}
                >
                  Clique aqui
                </button>{' '}
                para gerar um novo
              </>
            )}
          </div>
        )}
      </Alert>
    </AlertWrapper>
  )
}

export default EmailConfirmationAlert
