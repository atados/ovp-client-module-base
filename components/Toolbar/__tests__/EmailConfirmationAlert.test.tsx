import React from 'react'
import renderer from 'react-test-renderer'
import { render, fireEvent } from '@testing-library/react'
import EmailConfirmationAlert from '../EmailConfirmationAlert'
import moment from 'moment'

jest.mock('~/hooks/use-fetch-api-mutation', () => () => ({
  mutate: jest.fn(),
  loading: false,
  data: {
    success: false,
  },
}))

jest.mock('~/common', () => ({
  Config: {
    emailConfirmation: {
      warning: true,
    },
  },
}))

afterEach(() => {
  localStorage.clear()
})

describe('EmailConfirmationAlert', () => {
  const user = {
    uuid: '123456789',
    email: 'foo@bar.com',
    is_email_verified: false,
  }

  it('renders correctly', () => {
    localStorage.setItem(
      'hiddenEmailConfirmation',
      JSON.stringify({
        userUuid: user.uuid,
        expirationDate: moment()
          .subtract(1, 'week')
          .format(),
      }),
    )

    const renderedComponent = renderer.create(
      <EmailConfirmationAlert user={user} />,
    )
    renderedComponent.update(<EmailConfirmationAlert user={user} />)
    const tree = renderedComponent.toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('should send email and set hiddenEmailConfirmation on localstorage when button is pressed', () => {
    const root = render(<EmailConfirmationAlert user={user} />)

    fireEvent.click(root.getByRole('send-confirmation-email'))

    expect(
      JSON.parse(localStorage.getItem('hiddenEmailConfirmation') as string)
        .userUuid,
    ).toBe('123456789')
  })
})
