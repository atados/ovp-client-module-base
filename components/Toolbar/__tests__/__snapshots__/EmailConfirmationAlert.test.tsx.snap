// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmailConfirmationAlert renders correctly 1`] = `
.c0 {
  z-index: 51 !important;
}

<div
  className="fixed left-0 bottom-0 right-0 c0"
>
  <div
    className="w-full min-h-4 flex justify-center items-center py-2 px-4 bg-orange-400 bg-orange-400"
  >
    <div>
      <b
        className="mr-1"
      >
        Seu email ainda não foi confirmado.
      </b>
       Perdeu o email de confirmação?
       
      <button
        className="inline-block mx-1 underline"
        onClick={[Function]}
        role="send-confirmation-email"
      >
        Clique aqui
      </button>
       
      para gerar um novo
    </div>
    <button
      className="absolute right-0 mr-2 cursor-pointer"
      onClick={[Function]}
      role="close"
    >
      <span
        className="icon im"
      >
        close
      </span>
    </button>
  </div>
</div>
`;
